Add page reloading in offline.php templates

thelovekesh opened the pull request:
Fixes #438  

Automatically reload the offline page once the user is back online.

Tasks Done:

- [x] Add JS to automatically reload the offline page once the user is back online

https://user-images.githubusercontent.com/54371619/152957784-e01b9259-ccde-46f3-8f85-4d0404e7b436.mp4

thelovekesh committed:
Add page reloading in offline.php templates

* Automatically reload the offline page once the user is back online.

westonruter commented on wp-includes/template.php:
```diff
@@ -171,3 +171,47 @@ function wp_service_worker_error_details_template( $output = '' ) {
+		 */
+		async function checkNetworkAndReload() {
+			try {
+				const response = await fetch('.', {method: 'HEAD'});
```
Any reason for `.` as opposed to `location.href`? The problem with `.` is it will drop the query parameters. 

```suggestion
				const response = await fetch(location.href, {method: 'HEAD'});
```

thelovekesh replied:
That's interesting. Above we are making a fetch request to base origin to check whether we are getting response OK or not. I wonder how making a fetch request to some origin can drop query params of another origin.

westonruter commented on wp-includes/template.php:
```diff
@@ -171,3 +171,47 @@ function wp_service_worker_error_details_template( $output = '' ) {
+			} catch {
+				console.log('User is offline');
+			}
+			window.setTimeout(checkNetworkAndReload, 2500);
```
Any reason specifically for 2,500ms? I suppose just because this is what was used in https://github.com/GoogleChrome/web.dev/pull/4717.

If the server is having trouble due to heavy traffic or some other reason, getting hit every 2.5 seconds will add additional load and may be interpreted by the host as a DOS attack. Therefore, I suggest the value here implement [exponential backoff](https://en.wikipedia.org/wiki/Exponential_backoff) for retrying. So there could be a counter in the script for the number of tries, and each time the logic runs and the counter is incremented, the next timeout delay can be multiplied by the counter. 

Count | Delay
------|------
1 | 2.5s
2 | 5s
3 | 10s
4 | 20s
5 | 40s

In other words, multiple the 2500 delay by `Math.pow( 2, count )` where the initial count is `0`.

thelovekesh replied:
I think we should also reset the counter after the count is quals to 5 or 6. There can be a possibility that the user or server comes back online and page reloading has to wait for another 1 or 2 minutes due to time delay. What do you think?

ankitrox replied:
@thelovekesh We can cap the number of retries, so after `n` number of retries, there won't be more.

thelovekesh replied:
@westonruter Please let me know your thoughts on it.

westonruter replied:
I think the exponential backoff will be enough. After 15 tries the timeout goes up to ~1 day, so it will effectively be disabled once the count goes high enough.

westonruter commented on wp-includes/template.php:
```diff
@@ -171,3 +171,47 @@ function wp_service_worker_error_details_template( $output = '' ) {
+		return;
+	}
+	?>
+	<script>
```
Should this go through JS Coding Guidelines? https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/

Or should we rather apply Prettier formatting?

thelovekesh replied:
> apply Prettier formatting

~~Sounds good to me.~~

I think JS code in PHP files that follows WPCS should follow WP JS standards.

westonruter replied:
Note that core may be moving to Prettier: https://make.wordpress.org/core/2022/01/05/proposal-changes-to-javascript-coding-standards-for-full-prettier-compatibility/

westonruter commented on wp-includes/template.php:
```diff
@@ -171,3 +171,47 @@ function wp_service_worker_error_details_template( $output = '' ) {
+ * @since 0.7
+ */
+function wp_service_worker_offline_page_reload() {
+	if ( ! is_offline() ) {
```
It would be useful to also do the reload in the case of server being down.

Also, something else that comes to mind: what happens if the error page is served in response to a POST request? If so, each reload will cause a re-POST message from the browser and that's probably not desired.

```suggestion
	if ( ! is_offline() || ! is_500() || 'GET' !== $_SERVER['REQUEST_METHOD'] ) {
```

thelovekesh replied:
Thanks for pointing this out. For sure we need to check the method of the request. I have implemented these changes.

westonruter replied:
As part of #728, I just realized that this check will never do anything:

https://github.com/GoogleChromeLabs/pwa-wp/blob/124baabe30f70b718960fa18158f10a54943287e/wp-includes/template.php#L181-L184

This is because the offline template is always fetched via a `GET` request by the service worker when it is installing, so it will never not be `GET`. We need an alternative way of preventing the code from running on responses to non-GET requests.

Contrary to what I thought above, each reload does _not_ cause the browser re-POST message: https://github.com/GoogleChromeLabs/pwa-wp/pull/728#issuecomment-1099718487

westonruter replied:
I believe I've addressed this via https://github.com/GoogleChromeLabs/pwa-wp/pull/763

westonruter replied:
See https://github.com/GoogleChromeLabs/pwa-wp/pull/763

westonruter commented on wp-includes/template.php:
```diff
@@ -171,3 +171,47 @@ function wp_service_worker_error_details_template( $output = '' ) {
+			}
+			window.setTimeout(checkNetworkAndReload, 2500);
+		}
+		checkNetworkAndReload();
```
Either the entire script should be wrapped in an IIFE, or the `<script>` should have `type="module"` added to it. This will prevent the `checkNetworkAndReload` function from polluting the global scope.

thelovekesh committed:
Add counter to delay reloading in each retry

thelovekesh committed:
Update: Make fetch request to location.href instead of base origin

Making fetch requests to the base origin i.e. fetch('.') will drop the query params.

Co-authored-by: Weston Ruter <westonruter@google.com>

thelovekesh committed:
Add reloading script in offline and 500 templates and allow only on GET method

westonruter commented:
Note that testing this is a bit tricky because the script is part of the precached template, so without a version bump it won't show up.

westonruter commented:
Let's just add a PHPUnit test to ensure that the cases are covered for this function outputting the `script`.

thelovekesh committed:
Add output control function

thelovekesh committed:
Add test for wp_service_worker_offline_page_reload

thelovekesh commented:
> Note that testing this is a bit tricky because the script is part of the precached template, so without a version bump it won't show up.

@westonruter  Tests added 👍

westonruter commented on wp-includes/template.php:
```diff
@@ -171,3 +171,57 @@ function wp_service_worker_error_details_template( $output = '' ) {
+		checkNetworkAndReload();
+	</script>
+	<?php
+	return ob_get_clean();
```
Why adding the return here? If I'm not mistaken, the output buffering is causing the `script` to no longer be printed onto the page? The output buffering should rather be done in the tests.

thelovekesh replied:
> output buffering is causing the script to no longer be printed

@westonruter Ahh. my bad. Might get confused during the testing. 
Fixed this now 👍

thelovekesh committed:
Remove output control functions

thelovekesh committed:
Update test cases by adding output control function in output

thelovekesh commented:
@westonruter Can you please confirm if this PR needs no more improvements?

westonruter reviewed (approved):
This is good, yes.

thelovekesh commented:
> then the result is an infinite reload 
Thanks for pointing it out and it's a great finding. I just checked it after your comment. We should add it else it will cause infinite reloads during development.

westonruter commented:
@thelovekesh Yes please! Thanks.
